docs/manual/source/templates/recommendation/blacklist-items.html.md

```diff
@@ -0,0 +1,140 @@
+---
+title: Filter Recommended Items by Blacklist in Query (Recommendation)
+---
+
+Lets imagine that we want to avoid recommendations of items that user has already seen.
+We need to slightly modify Recommendation template
+
+## Add Query Parameter
+
+First of all we need to specify query parameter to send items ids that the user has already seen.
+Lets modify `case class Query` in MyRecommendation/src/main/scala/***Engine.scala***:
+
+```scala
+case class Query(
+  user: String,
+  num: Int,
+  blackList: Set[String] // NOTE: line added
+) extends Serializable
+```
+
+## Filter the Data
+
+Then we need to change the code that computes recommendation score to filter out the seen items.
+Lets modify class MyRecommendation/src/main/scala/***ALSModel.scala***. Just add the following two methods to that class.
+
+```scala
+  def recommendProductsWithFilter(
+      user: Int,
+      num: Int,
+      productIdFilter: Set[Int]) = {
+    val filteredProductFeatures = productFeatures
+      .filter(features => !productIdFilter.contains(features._1)) // (*)
+    recommend(userFeatures.lookup(user).head, filteredProductFeatures, num)
+      .map(t => Rating(user, t._1, t._2))
+  }
+
+  private def recommend(
+      recommendToFeatures: Array[Double],
+      recommendableFeatures: RDD[(Int, Array[Double])],
+      num: Int): Array[(Int, Double)] = {
+    val recommendToVector = new DoubleMatrix(recommendToFeatures)
+    val scored = recommendableFeatures.map { case (id,features) =>
+      (id, recommendToVector.dot(new DoubleMatrix(features)))
+    }
+    scored.top(num)(Ordering.by(_._2))
+  }
+```
+
+Also it’s required to add import of the `org.jblas.DoubleMatrix` class.
+Please make attention that method `recommend` is the copy of method `org.apache.spark.mllib.recommendation.MatrixFactorizationModel#recommend`. We can't reuse this because it’s private.
+Method `recommendProductsWithFilter` is the almost full copy of `org.apache.spark.mllib.recommendation.MatrixFactorizationModel#recommendProduct` method. The difference only is the line with commentary ‘(*)’ where we apply filtering.
+
+## Put It All Together
+
+Next we need to invoke our new method with filtering when we query recommendations.
+Lets modify method `predict` in MyRecommendation/src/main/scala/***ALSAlgorithm.scala***:
+
+```scala
+  def predict(model: ALSModel, query: Query): PredictedResult = {
+    // Convert String ID to Int index for Mllib
+    model.userStringIntMap.get(query.user).map { userInt =>
+      // create inverse view of itemStringIntMap
+      val itemIntStringMap = model.itemStringIntMap.inverse
+      // recommendProducts() returns Array[MLlibRating], which uses item Int
+      // index. Convert it to String ID for returning PredictedResult
+      val blackListedIds = query.blackList.map(model.itemStringIntMap.get(_).get) // NOTE: line added
```

Comment:
Hi @aparamonov , need to handle the case that the blackListed item IDs not found in model (such as new item ID). eg.
val blackListedIds = query.blackList.map(model.itemStringIntMap.get(_)).flatten